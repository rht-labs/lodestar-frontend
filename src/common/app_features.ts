// All roles defined by the backend to be used in the frontend should be enumerated in this object.

export const APP_FEATURES = {
  admin: 'admin',
  writer: 'writer',
  reader: 'reader',
  notifications: 'notifications',
  engagementCardIcons: 'engagementCardIcons',
  resetUser: 'resetUser',
};

export type AppFeature = keyof typeof APP_FEATURES;
