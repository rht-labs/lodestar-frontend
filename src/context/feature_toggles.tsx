import React, { useContext } from 'react';
import { SessionContext } from './session_context';

interface FeatureToggleContext {
  features: string[];
}

export const FeatureToggleContext = React.createContext<FeatureToggleContext>({
  features: [],
});

export const FeatureToggles = ({ children }: { children: React.ReactNode }) => {
  const sessionContext = useContext(SessionContext);
  const features = sessionContext.sessionData?.roles ?? [];
  return (
    <FeatureToggleContext.Provider value={{ features }}>
      {children}
    </FeatureToggleContext.Provider>
  );
};
