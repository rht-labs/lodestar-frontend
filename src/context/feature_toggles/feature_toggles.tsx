import React, { useCallback } from 'react';
import { useSession } from '../auth_context/auth_context';

interface FeatureToggleContext {
  features: string[];
  hasFeature: (name: string) => boolean;
}

export const FeatureToggleContext = React.createContext<FeatureToggleContext>({
  features: [],
  hasFeature: () => false,
});

export const FeatureToggles = ({
  children,
  features,
}: {
  children: React.ReactNode;
  features?: string[];
}) => {
  const authContext = useSession();
  const providedFeatures = features ?? authContext.sessionData?.roles ?? [];
  const hasFeature = useCallback(
    (name: string) => {
      return name && providedFeatures.includes(name);
    },
    [providedFeatures]
  );

  return (
    <FeatureToggleContext.Provider
      value={{ features: providedFeatures, hasFeature }}
    >
      {children}
    </FeatureToggleContext.Provider>
  );
};
