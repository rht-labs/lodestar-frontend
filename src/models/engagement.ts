export interface Engagement {
  name: string;
}
export class Engagement {
  constructor({ name }: Engagement) {
    this.name = name;
  }
  name: string;
}
