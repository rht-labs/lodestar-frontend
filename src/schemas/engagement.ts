import faker from 'faker/locale/en_US';
import { LaunchData } from './launch_data';
import { GitCommit } from './git_commit';
import { CreationDetails } from './creation_details';
import { ClusterStatus } from './cluster_status';
export interface Engagement {
  additional_details?: string;
  archive_date: Date;
  commits: GitCommit[];
  customer_contact_email: string;
  customer_contact_name: string;
  customer_name: string;
  description: string;
  end_date: Date;
  engagement_lead_email: string;
  engagement_lead_name: string;
  engagement_users: Array<any>;
  last_update: string;
  location: string;
  mongo_id?: string;
  ocp_cloud_provider_name: string;
  ocp_cloud_provider_region: string;
  ocp_cluster_size: string;
  ocp_persistent_storage_size: string;
  ocp_sub_domain: string;
  ocp_version: string;
  project_id: number;
  project_name: string;
  start_date: Date;
  technical_lead_email: string;
  technical_lead_name: string;
  launch?: LaunchData;
  creation_details: CreationDetails;
  last_update_by_name: string;
  suggested_subdomain?: string;
  status: ClusterStatus;
}
export abstract class Engagement {
  static fromFake(staticData = false): Engagement {
    return {
      additional_details: staticData
        ? 'Additional information here'
        : faker.lorem.paragraphs(2),
      archive_date: staticData ? new Date(2020, 6, 30) : faker.date.recent(),
      commits: [],
      customer_contact_email: staticData
        ? 'bob@doe.com'
        : faker.internet.email(),
      customer_contact_name: staticData
        ? 'Bob Doe'
        : `${faker.name.firstName()} ${faker.name.lastName()}`,
      customer_name: staticData ? 'NASA' : faker.company.companyName(),
      description: staticData
        ? "It's rocket science"
        : faker.lorem.paragraphs(5),
      end_date: staticData ? new Date(2020, 6, 1) : faker.date.future(),
      engagement_users: staticData ? [] : [],
      engagement_lead_email: staticData
        ? 'alice@doe.com'
        : faker.internet.email(),
      engagement_lead_name: staticData
        ? 'Alice Doe'
        : `${faker.name.firstName()} ${faker.name.lastName()}`,
      last_update: staticData ? '2020-01-01' : new Date().toString(),
      location: staticData
        ? 'Nashville, TN'
        : `${faker.address.city()}, ${faker.address.stateAbbr()}`,
      mongo_id: '1',
      ocp_cloud_provider_name: staticData ? 'AWS' : 'AWS',
      ocp_cloud_provider_region: staticData
        ? 'N. Virginia'
        : faker.lorem.word(),
      ocp_cluster_size: staticData ? 'Large' : 'Large',
      ocp_persistent_storage_size: staticData
        ? '100G'
        : `${faker.random.number()}G`,
      ocp_sub_domain: staticData ? 'ordoabchao' : faker.lorem.word(),
      ocp_version: staticData ? '4.4' : faker.random.number().toString(),
      project_id: staticData ? 1 : faker.random.number(),
      project_name: staticData ? 'Boots on the Moon' : faker.company.bsNoun(),
      start_date: staticData ? new Date(2020, 5, 1) : faker.date.recent(),
      technical_lead_email: staticData ? 'eve@doe.com' : faker.internet.email(),
      technical_lead_name: staticData
        ? 'Eve Doe'
        : `${faker.name.firstName()} ${faker.name.lastName()}`,
      creation_details: staticData
        ? {
            created_by_email: 'dwasinger@redhat.com',
            created_by_user: 'dwasinge',
            created_on: new Date(2020, 1, 1),
          }
        : {
            created_by_email: faker.internet.email(),
            created_by_user: `${faker.name.firstName()} ${faker.name.lastName()}`,
            created_on: faker.date.recent(),
          },
      last_update_by_name: staticData
        ? 'James Doe'
        : `${faker.name.firstName()} ${faker.name.lastName()}`,
      launch: staticData
        ? null
        : faker.random.boolean()
        ? {
            launched_by: faker.name.firstName(),
            launched_date_time: faker.date.recent(),
          }
        : null,
      status: ClusterStatus.fromFake(staticData),
    };
  }
}

export enum EngagementStatus {
  active = 'active',
  past = 'past',
  upcoming = 'upcoming',
}

export const getEngagementStatus = (
  engagement: Engagement
): EngagementStatus => {
  if (!engagement) {
    return null;
  }
  const Today = new Date();
  const { launch, end_date } = engagement;
  const hasLaunched = !!launch?.launched_date_time;
  if (hasLaunched && end_date >= Today) {
    return EngagementStatus.active;
  } else if (hasLaunched && end_date < Today) {
    return EngagementStatus.past;
  } else {
    return EngagementStatus.upcoming;
  }
};
