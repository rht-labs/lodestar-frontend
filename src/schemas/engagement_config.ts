import FakedSchema from './fixtures/engagement_form_config.json';

export interface EngagementFormOption {
  label: string;
  value: string;
  options?: EngagementFormOption[];
}

export interface EngagementFormValidator {
  kind: string;
  value?: any;
}

export interface FormConfig {
  options?: EngagementFormOption[];
  validators?: EngagementFormValidator[];
}

export interface EngagementFormConfig {
  [key: string]: { [key: string]: FormConfig };
}

export abstract class EngagementFormConfig {
  static fromFake(): EngagementFormConfig {
    return FakedSchema;
  }
}
