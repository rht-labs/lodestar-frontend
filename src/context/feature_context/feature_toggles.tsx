import React, { useCallback, useEffect, useState } from 'react';
import { useLocation } from 'react-router';
import {
  getFeaturesFromVersion,
  FEATURE_VERSION_MAP,
} from '../../common/version_feature_factory';
import { IAuthContext } from '../auth_context/auth_context';
import { IVersionContext } from '../version_context/version_context';
import qs from 'query-string';
import { Config } from '../../schemas/config';

export interface IFeatureToggleContext {
  features: string[];
  isLoading: boolean;
  hasFeature: (name: string) => boolean;
}

export const FeatureToggleContext = React.createContext<IFeatureToggleContext>({
  features: [],
  hasFeature: () => false,
  isLoading: false,
});

export interface IFeatureToggleProvider {
  features?: string[];
  authContext: IAuthContext;
  versionContext: IVersionContext;
  config?: Config;
  children: any;
}

export const FeatureToggles = ({
  children,
  features,
  authContext,
  versionContext,
  config,
}: IFeatureToggleProvider) => {
  let version = versionContext?.versions?.mainVersion?.value;

  const location = useLocation();
  if (config?.allowVersionOverride) {
    const query = qs.parse(location.search);
    const queryVersion = query['lodestar-version'];
    if (queryVersion) {
      version = Array.isArray(queryVersion) ? queryVersion[0] : queryVersion;
    }
  }
  const versionFeatures = getFeaturesFromVersion(version, FEATURE_VERSION_MAP);
  const roleFeatures = (authContext?.sessionData?.roles ?? []).concat(
    features ?? []
  );
  const [hasFetchedVersions, setHasFetchedVersions] = useState(false);
  useEffect(() => {
    if (!hasFetchedVersions) {
      versionContext?.fetchVersions?.();
      setHasFetchedVersions(true);
    }
  }, [versionContext, hasFetchedVersions]);
  const allFeatures = React.useMemo(
    () => [...versionFeatures, ...roleFeatures],
    [versionFeatures, roleFeatures]
  );
  const hasFeature = useCallback(
    (name: string) => {
      return name && allFeatures.includes(name);
    },
    [allFeatures]
  );
  const isLoading = !versionContext?.versions;

  return (
    <FeatureToggleContext.Provider
      value={{ features: allFeatures, hasFeature, isLoading }}
    >
      {children}
    </FeatureToggleContext.Provider>
  );
};
