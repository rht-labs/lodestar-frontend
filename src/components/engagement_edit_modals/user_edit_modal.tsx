import React from 'react';
import {
  Button,
  EmptyStateIcon,
  Form,
  FormSelect,
  FormSelectOption,
  InputGroup,
  Text,
  TextInput,
  EmptyState,
  Title,
  EmptyStateBody,
  Modal,
  ModalVariant,
  Grid,
  GridItem,
} from '@patternfly/react-core';

import {
  PlusCircleIcon,
  ErrorCircleOIcon,
  UsersIcon,
} from '@patternfly/react-icons';
import { Feature } from '../../components/feature';
import { APP_FEATURES } from '../../common/app_features';
import { Engagement } from '../../schemas/engagement_schema';
import { EngagementFormConfig } from '../../schemas/engagement_config';
import { useFeatures } from '../../context/feature_toggles/feature_hook';
import { useModalVisibility } from '../../context/edit_modal_visibility_context/edit_modal_visibility_hook';
import { EditModalTemplate } from '../../layout/edit_modal_template';

export interface UserEditModalProps {
  onChange: (fieldName: string, value: any) => void;
  formOptions: EngagementFormConfig;
  engagement: Engagement;
  isOpen: boolean;
  onSave: (engagement: Engagement) => void;
}
export function UserEditModal({
  engagement,
  formOptions,
  onChange,
  isOpen,
  onSave: propsOnSave,
}: UserEditModalProps) {
  const { requestClose } = useModalVisibility();
  const { hasFeature } = useFeatures();
  const onSave = () => {
    propsOnSave(engagement);
    requestClose();
  };
  function addUser() {
    const newUser = { first_name: '', last_name: '', email: '', role: '' };
    engagement.engagement_users.push(newUser);
    onChange('user', engagement.engagement_users);
  }
  function removeUser(index: any) {
    engagement.engagement_users.splice(index.currentTarget.value, 1);
    onChange('user', engagement.engagement_users);
  }
  return (
    <Modal
      variant={ModalVariant.large}
      isOpen={isOpen}
      onClose={requestClose}
      title="Engagement Users"
    >
      <EditModalTemplate
        actions={
          <div>
            <Button onClick={onSave} data-cy={'save_users'} >Save</Button>
          </div>
        }
      >
        <div>
          {!engagement.engagement_users.length ? (
            <EmptyState>
              <EmptyStateIcon icon={UsersIcon} />
              <Title headingLevel="h4" size="lg">
                No Users Added
              </Title>
              <EmptyStateBody>
                <p>No users have been added to this engagement</p>
                <p>
                  Select the 'add user' button below, to begin adding users.
                </p>
              </EmptyStateBody>
              <Button variant="primary" onClick={addUser} data-cy={'add_new_user'}>
                Add User
              </Button>
            </EmptyState>
          ) : (
            <div>
              <Grid>
                <Form isHorizontal>
                  <ul>
                    <li>
                      <InputGroup>
                        <Grid style={{ width: '100%' }}>
                          <GridItem span={3}>
                            <Text>First Name</Text>
                          </GridItem>
                          <GridItem span={3}>
                            <Text>Last Name</Text>
                          </GridItem>
                          <GridItem span={3}>
                            <Text>Email</Text>
                          </GridItem>
                          <GridItem span={2}>
                            <Text>Role</Text>
                          </GridItem>
                          <GridItem span={1}>
                            <Feature name={APP_FEATURES.writer}>
                              <Text>Del</Text>
                            </Feature>
                          </GridItem>
                        </Grid>
                      </InputGroup>
                    </li>
                    {engagement.engagement_users.map(
                      (value: any, index: any) => {
                        return (
                          <li key={index}>
                            <InputGroup>
                              <Grid style={{ width: '100%' }}>
                                <GridItem span={3}>
                                  <TextInput
                                    aria-label="First Name"
                                    name="first-name"
                                    data-cy={'input_user_firstname'}
                                    isDisabled={
                                      !hasFeature(APP_FEATURES.writer)
                                    }
                                    onChange={e => {
                                      engagement.engagement_users[
                                        index
                                      ].first_name = e;
                                      onChange(
                                        'user',
                                        engagement.engagement_users
                                      );
                                    }}
                                    placeholder="First Name"
                                    type="text"
                                    value={value.first_name || ''}
                                  />
                                </GridItem>
                                <GridItem span={3}>
                                  <TextInput
                                    aria-label="Last Name"
                                    name="last-name"
                                    data-cy={'input_user_lastname'}
                                    isDisabled={
                                      !hasFeature(APP_FEATURES.writer)
                                    }
                                    onChange={e => {
                                      engagement.engagement_users[
                                        index
                                      ].last_name = e;
                                      onChange(
                                        'user',
                                        engagement.engagement_users
                                      );
                                    }}
                                    placeholder="Last Name"
                                    type="text"
                                    value={value.last_name || ''}
                                  />
                                </GridItem>
                                <GridItem span={3}>
                                  <TextInput
                                    aria-label="email"
                                    name="email"
                                    data-cy={'input_user_email'}
                                    isDisabled={
                                      !hasFeature(APP_FEATURES.writer)
                                    }
                                    onChange={e => {
                                      engagement.engagement_users[
                                        index
                                      ].email = e;
                                      onChange(
                                        'user',
                                        engagement.engagement_users
                                      );
                                    }}
                                    placeholder="Email Address"
                                    type="email"
                                    value={value.email || ''}
                                  />
                                </GridItem>
                                <GridItem span={2}>
                                  <FormSelect
                                    name="role"
                                    aria-label="User Role"
                                    id='user_role_dropdown'
                                    value={value.role || ''}
                                    isDisabled={
                                      !hasFeature(APP_FEATURES.writer)
                                    }
                                    onChange={e => {
                                      engagement.engagement_users[
                                        index
                                      ].role = e;
                                      onChange(
                                        'user',
                                        engagement.engagement_users
                                      );
                                    }}
                                  >
                                    {[
                                      <FormSelectOption
                                        isDisabled={true}
                                        key={'placeholder'}
                                        value={undefined}
                                        label={'Select a role'}
                                      />,
                                    ].concat(
                                      (
                                        formOptions?.user_options?.user_roles
                                          ?.options ?? []
                                      )?.map((option: any, index: number) => (
                                        <FormSelectOption
                                          isDisabled={option.disabled}
                                          key={index}
                                          value={option.value}
                                          label={option.label}
                                          data-cy={option.label}
                                        />
                                      ))
                                    )}
                                  </FormSelect>
                                </GridItem>
                                <GridItem span={1}>
                                  <Feature name={APP_FEATURES.writer}>
                                    <Button
                                      onClick={removeUser}
                                      value={index}
                                      variant="danger"
                                      isInline
                                    >
                                      <ErrorCircleOIcon />
                                    </Button>
                                  </Feature>
                                </GridItem>
                              </Grid>
                            </InputGroup>
                          </li>
                        );
                      }
                    )}
                  </ul>
                </Form>
              </Grid>
              <Feature name={APP_FEATURES.writer}>
                <Button
                  onClick={addUser}
                  variant="link"
                  icon={<PlusCircleIcon />}
                  data-cy={'add_more_users'}
                >
                  Add User
                </Button>
              </Feature>
            </div>
          )}
        </div>
      </EditModalTemplate>
    </Modal>
  );
}
