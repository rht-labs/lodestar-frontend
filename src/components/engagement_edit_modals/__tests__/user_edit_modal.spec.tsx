import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { UserEditModal } from '../user_edit_modal';
import { Engagement } from '../../../schemas/engagement';
import { EngagementFormConfig } from '../../../schemas/engagement_config';
import { TestStateWrapper } from '../../../common/test_state_wrapper';
import { FeatureToggleContext } from '../../../context/feature_context/feature_toggles';

describe('Point of Contact edit modal', () => {
  test('matches snapshot', () => {
    expect(
      render(
        <UserEditModal
          onSave={() => {}}
          formOptions={EngagementFormConfig.fromFake()}
          isOpen={true}
          engagement={Engagement.fromFake(true)}
          onChange={() => {}}
        />
      )
    ).toMatchSnapshot();
  });

  test('When clicking the save button, the onSave method is called', async () => {
    const onSave = jest.fn();
    const { getByTestId } = render(
      <UserEditModal
        onSave={onSave}
        engagement={Engagement.fromFake(true)}
        formOptions={EngagementFormConfig.fromFake()}
        isOpen={true}
        onChange={() => {}}
      />
    );
    await fireEvent.click(getByTestId('user-edit-save'));
    expect(onSave).toHaveBeenCalled();
  });
  test('Clicking the addUser button calls onChange', async () => {
    const onChange = jest.fn();
    const { getByTestId } = render(
      <FeatureToggleContext.Provider
        value={{ features: ['reader', 'writer'], hasFeature: () => true }}
      >
        <UserEditModal
          onSave={() => {}}
          engagement={({ engagement_users: [] } as unknown) as Engagement}
          formOptions={EngagementFormConfig.fromFake()}
          isOpen={true}
          onChange={onChange}
        />
      </FeatureToggleContext.Provider>
    );
    await fireEvent.click(getByTestId('add-first-user'));
    expect(onChange).toHaveBeenCalledWith('user', [
      {
        email: '',
        first_name: '',
        last_name: '',
        role: '',
      },
    ]);
  });
  test('Clicking the removeUser button calls onChange', async () => {
    const onChange = jest.fn();
    const { getByTestId } = render(
      <FeatureToggleContext.Provider
        value={{ features: ['reader', 'writer'], hasFeature: () => true }}
      >
        <UserEditModal
          onSave={() => {}}
          engagement={Engagement.fromFake(true)}
          formOptions={EngagementFormConfig.fromFake()}
          isOpen={true}
          onChange={onChange}
        />
      </FeatureToggleContext.Provider>
    );
    await fireEvent.click(getByTestId('remove-user-button-0'));
    expect(onChange).toHaveBeenCalledWith('user', []);
  });
});
