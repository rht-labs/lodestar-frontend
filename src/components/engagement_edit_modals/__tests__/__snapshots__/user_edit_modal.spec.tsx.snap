// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Point of Contact edit modal matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    class="pf-c-backdrop__open"
  >
    <div
      aria-hidden="true"
    />
    <div>
      <div
        class="pf-c-backdrop"
      >
        <div
          class="pf-l-bullseye"
        >
          <div
            aria-describedby="pf-modal-part-2"
            aria-labelledby="pf-modal-part-1"
            aria-modal="true"
            class="pf-c-modal-box pf-m-lg"
            id="pf-modal-part-0"
            role="dialog"
          >
            <button
              aria-label="Close"
              class="pf-c-button pf-m-plain"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe="true"
              type="button"
            >
              <svg
                aria-hidden="true"
                fill="currentColor"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 352 512"
                width="1em"
              >
                <path
                  d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                  transform=""
                />
              </svg>
            </button>
            <header
              class="pf-c-modal-box__header"
            >
              <h1
                class="pf-c-modal-box__title"
                id="pf-modal-part-1"
              >
                Engagement Users
              </h1>
            </header>
            <div
              class="pf-c-modal-box__body"
              id="pf-modal-part-2"
            >
              <section
                class="pf-c-page__main-section"
              >
                <div>
                  <div>
                    <div
                      class="pf-l-grid"
                    >
                      <form
                        class="pf-c-form pf-m-horizontal"
                        novalidate=""
                      >
                        <ul>
                          <li>
                            <div
                              class="pf-c-input-group"
                            >
                              <div
                                class="pf-l-grid"
                                style="width: 100%;"
                              >
                                <div
                                  class="pf-l-grid__item pf-m-3-col"
                                >
                                  <p
                                    class=""
                                    data-pf-content="true"
                                  >
                                    First Name
                                  </p>
                                </div>
                                <div
                                  class="pf-l-grid__item pf-m-3-col"
                                >
                                  <p
                                    class=""
                                    data-pf-content="true"
                                  >
                                    Last Name
                                  </p>
                                </div>
                                <div
                                  class="pf-l-grid__item pf-m-3-col"
                                >
                                  <p
                                    class=""
                                    data-pf-content="true"
                                  >
                                    Email
                                  </p>
                                </div>
                                <div
                                  class="pf-l-grid__item pf-m-2-col"
                                >
                                  <p
                                    class=""
                                    data-pf-content="true"
                                  >
                                    Role
                                  </p>
                                </div>
                                <div
                                  class="pf-l-grid__item pf-m-1-col"
                                />
                              </div>
                            </div>
                          </li>
                          <li>
                            <div
                              class="pf-c-input-group"
                            >
                              <div
                                class="pf-l-grid"
                                style="width: 100%;"
                              >
                                <div
                                  class="pf-l-grid__item pf-m-3-col"
                                >
                                  <input
                                    aria-invalid="false"
                                    aria-label="First Name"
                                    class="pf-c-form-control"
                                    data-cy="input_user_firstname"
                                    disabled=""
                                    name="first-name"
                                    placeholder="First Name"
                                    type="text"
                                    value="John"
                                  />
                                </div>
                                <div
                                  class="pf-l-grid__item pf-m-3-col"
                                >
                                  <input
                                    aria-invalid="false"
                                    aria-label="Last Name"
                                    class="pf-c-form-control"
                                    data-cy="input_user_lastname"
                                    disabled=""
                                    name="last-name"
                                    placeholder="Last Name"
                                    type="text"
                                    value="Doe"
                                  />
                                </div>
                                <div
                                  class="pf-l-grid__item pf-m-3-col"
                                >
                                  <input
                                    aria-invalid="false"
                                    aria-label="email"
                                    class="pf-c-form-control"
                                    data-cy="input_user_email"
                                    disabled=""
                                    name="email"
                                    placeholder="Email Address"
                                    type="email"
                                    value="jdoe@doe.co"
                                  />
                                </div>
                                <div
                                  class="pf-l-grid__item pf-m-2-col"
                                >
                                  <select
                                    aria-invalid="false"
                                    aria-label="User Role"
                                    class="pf-c-form-control"
                                    disabled=""
                                    id="user_role_dropdown"
                                    name="role"
                                  >
                                    <option
                                      class=""
                                      disabled=""
                                      value=""
                                    >
                                      Select a role
                                    </option>
                                  </select>
                                </div>
                                <div
                                  class="pf-l-grid__item pf-m-1-col"
                                />
                              </div>
                            </div>
                          </li>
                        </ul>
                      </form>
                    </div>
                  </div>
                </div>
              </section>
              <section
                class="pf-c-page__main-section"
              >
                <div
                  style="margin-top: 1rem;"
                >
                  <div
                    class="pf-l-flex pf-m-justify-content-flex-end"
                  >
                    <div
                      class=""
                    >
                      <div>
                        <button
                          class="pf-c-button pf-m-primary"
                          data-cy="save_users"
                          data-ouia-component-type="PF4/Button"
                          data-ouia-safe="true"
                          data-testid="user-edit-save"
                          type="button"
                        >
                          Save
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </section>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
  />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
